### DISCUSSION TITLE
Try to add device, supports endpoints: but need to change the *topic* per endpoint (L&S Emotion)

### DISCUSSION THREAD

**Discussion by mslotema** [Koenkk/zigbee2mqtt#7373] [Devices]
Hi,

I am trying to add a new device to control the LED strips in my kitchen: two strips, powered and controlled separately.
And I run into a problem.

I have created a js file to support it, based on examples I found. And it works... sort of.

Well, it actually works 100%, but not in the zigbee2mqtt interface. Because it doesn't let me vary the topic per endpoint.
The device works by adding a **zone** (numerical value) in the **topic**:

```
    zigbee2mqtt/[name]/[zone]/set
```

It is the zone thing. That is a number. Without it it defaults to "1"; that works too. If I put a "2" there, the other side responds. Very nice :-)

_I can't find a way to define the device, which shows this particular behavior on the topic!_ 

(I might be missing something).

Let me explain in more detail.

I have bought a zigbee to 433MHz bridge for the LED lighting in my new kitchen. The kitchen company has no clue what to do
with this, which would be really funny, if it wasn't so expensive.
Anyhow, I went to work on it.

Of course, I want to use it with zigbee2mqtt (excellent stuff, btw!) and Home Assistant.

And I also want to add it to the codebase of zigbee2mqtt/herdsman. Of course!

The setup is the following:

- The brand is "L&S Deutschland GmbH", model "Emotion"
- I have two powerblocks to power the LEDs, left side and right side of the kitchen
- Each of which pair with separate 433MHz remote controls (two simple remotes)
- The zigbee gizmo is (I know now) capable of pairing with both blocks (in the lingo of L&S it is called a "zone"); I paired both sides
- The zigbee2mqtt interview shows 2 "endpoints". I suspect it support up to 4 "zones" or endpoints, because their top model remote supports that too.
- The js file I created only controls one side out of the box.
- With an MQTT tool, I can tweak the topic manually, and I can control both sides.

So, I need two things, besides the supported commands:

- I need to get zigbee2mqtt change the topic per endpoint
- I want to interview in the definition the number of endpoints it tells me - because this can be dynamic.

I tried to add endpoints in the js file, but this only adds the enpoint name to the commands (like: endpoint "1" results in 
command, e.g., "state_1", etc.). But this is not what is needed!

The topic needs to be changed.

And, I happen to have only two blocks. Other people might have more, or less. So the Z2M UI should reflect the number of registered zones.

I can't find in the sources if there is something implemented like this. And I also cannot guess what to do.
I might be overlooking something.

_Am I missing an option?_

This is my device definition:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const e = exposes.presets;

const device = {
    zigbeeModel: ['Emotion'],
    model: 'Emotion',
    vendor: 'LS Deutschland GmbH',
    description: 'LS Emotion zigbee bridge for LED controllers',
    fromZigbee: [fz.on_off, fz.brightness, fz.color_colortemp],
    toZigbee: [tz.light_onoff_brightness, tz.ignore_transition, tz.ignore_rate, tz.effect, 
        tz.light_brightness_move, tz.light_brightness_step, tz.light_colortemp, 
        tz.light_colortemp_move, tz.light_colortemp_step],
    meta: {disableDefaultResponse: true, multiEndpoint: false},
    exposes: [
        e.light_brightness_colortemp().removeFeature('color_temp_startup'),
    ],
};

module.exports = device;
```

The interview clearly shows endpoints, but it seems that they behave differently from endpoints that zigbee2mqtt normally uses:
 
```json
{
    "date_code": "",
    "definition": {
      "description": "LS Emotion zigbee controller",
      "exposes": [
        {
          "features": [
            {
              "access": 7,
              "description": "On/off state of this light",
              "name": "state",
              "property": "state",
              "type": "binary",
              "value_off": "OFF",
              "value_on": "ON",
              "value_toggle": "TOGGLE"
            },
            {
              "access": 7,
              "description": "Brightness of this light",
              "name": "brightness",
              "property": "brightness",
              "type": "numeric",
              "value_max": 254,
              "value_min": 0
            },
            {
              "access": 7,
              "description": "Color temperature of this light",
              "name": "color_temp",
              "presets": [
                {
                  "description": "Coolest temperature supported",
                  "name": "coolest",
                  "value": 150
                },
                {
                  "description": "Cool temperature (250 mireds / 4000 Kelvin)",
                  "name": "cool",
                  "value": 250
                },
                {
                  "description": "Neutral temperature (370 mireds / 2700 Kelvin)",
                  "name": "neutral",
                  "value": 370
                },
                {
                  "description": "Warm temperature (454 mireds / 2200 Kelvin)",
                  "name": "warm",
                  "value": 454
                },
                {
                  "description": "Warmest temperature supported",
                  "name": "warmest",
                  "value": 500
                }
              ],
              "property": "color_temp",
              "type": "numeric",
              "unit": "mired",
              "value_max": 500,
              "value_min": 150
            }
          ],
          "type": "light"
        }
      ],
      "model": "Emotion",
      "supports_ota": false,
      "vendor": "LS Deutschland GmbH"
    },
    "endpoints": {
      "1": {
        "bindings": [],
        "clusters": {
          "input": [
            "genBasic",
            "genIdentify",
            "genGroups",
            "genScenes",
            "genOnOff",
            "genLevelCtrl",
            "lightingColorCtrl"
          ],
          "output": []
        },
        "configured_reportings": []
      },
      "2": {
        "bindings": [],
        "clusters": {
          "input": [
            "genBasic",
            "genIdentify",
            "genGroups",
            "genScenes",
            "genOnOff",
            "genLevelCtrl",
            "lightingColorCtrl"
          ],
          "output": []
        },
        "configured_reportings": []
      },
      "242": {
        "bindings": [],
        "clusters": { "input": [], "output": ["greenPower"] },
        "configured_reportings": []
      }
    },
    "friendly_name": "<hex#>",
    "ieee_address": "<hex#>",
    "interview_completed": true,
    "interviewing": false,
    "model_id": "Emotion",
    "network_address": 14277,
    "power_source": "DC Source",
    "software_build_id": "1.0.0_r00000",
    "supported": true,
    "type": "Router"
  }

```

Now, again: the device works, on both the powerblocks. 
And the js-definition *seems* to work too, except you have to tweak the topic manually to get to both sides.

Is there something I missed?

Kind regards,

Melle

**Comment by Koenkk** [comment]
You can take this device as an example: https://github.com/Koenkk/zigbee-herdsman-converters/blob/4017685c6e9536cb96cd87f4756537d142e9b207/devices/gledopto.js#L171

**Reply by mslotema** [reply]
Hi Koen,

Thanks for responding.

I think you want me to look at the endpoint definitions. I tried that, in various formations:

```javascript
const device = {
    zigbeeModel: ['Emotion'],
    model: 'Emotion',
    vendor: 'LS Deutschland GmbH',
    description: 'LS Emotion zigbee bridge for LED controllers',
    fromZigbee: [fz.on_off, fz.brightness, fz.color_colortemp],
    toZigbee: [tz.light_onoff_brightness, tz.ignore_transition, tz.ignore_rate, tz.effect,
        tz.light_brightness_move, tz.light_brightness_step, tz.light_colortemp,
        tz.light_colortemp_move, tz.light_colortemp_step],
    meta: {disableDefaultResponse: true, multiEndpoint: false},
    exposes: [
        e.light_brightness_colortemp().removeFeature('color_temp_startup').withEndpoint(1),
        e.light_brightness_colortemp().removeFeature('color_temp_startup').withEndpoint(2),
        // and this should be dynamic ... when the device works
    ],
    endpoint: (device) => {
        return {'1': 1, '2': 2};
    },
};
```

Now, this indeed correctly changes the topic (missed that one the other night, apologies), but it also adds the endpoint name in the command that is being sent. And that is not working.

This is the result: 

```log
Zigbee2MQTT:debug 2021-05-07 20:10:16: Received MQTT message on 'zigbee2mqtt/0xccccccfffe6fb047/1/set' with data '{"state_1":"ON"}'
Zigbee2MQTT:error 2021-05-07 20:10:16: No converter available for 'state_1' (ON)
Zigbee2MQTT:debug 2021-05-07 20:10:20: Received MQTT message on 'zigbee2mqtt/0xccccccfffe6fb047/1/set' with data '{"state_1":"OFF"}'
Zigbee2MQTT:error 2021-05-07 20:10:20: No converter available for 'state_1' (OFF)
Zigbee2MQTT:debug 2021-05-07 20:10:22: Received MQTT message on 'zigbee2mqtt/0xccccccfffe6fb047/2/set' with data '{"state_2":"ON"}'
Zigbee2MQTT:error 2021-05-07 20:10:22: No converter available for 'state_2' (ON)
```
It is the ```state_1``` that I want to prevent.

The device expects this:

```log
Received MQTT message on 'zigbee2mqtt/0xccccccfffe6fb047/1/set' with data '{"state":"ON"}'
```

Am I still missing something?

Kind regards,
Melle

**Reply by Koenkk** [reply]
Only endpoint names in https://github.com/Koenkk/zigbee2mqtt/blob/f2e39af12cfaf46166b053ee43b28fb862d46731/lib/util/utils.js#L34 are allowed so try with e.g. `return {'l1': 1, 'l2': 2};`

**Reply by mslotema** [reply]
Hi,

I saw that too (and I tried it). 
But the thing is, the application adds the endpoint name in  the payload as well, on the data fields. 
I just need the endpoint to manipulate the topic. 

Maybe that is not yet supported. Then I know what to do

I would opt to add that, and call it something else. Channel, or something like that.

**Reply by Koenkk** [reply]
> But the thing is, the application adds the endpoint name in the payload as well, on the data fields.

Why is this a problem?

You have to see the endpoints as separate devices.

**Comment by mslotema** [comment]
Yes. I get that. 
But the device cannot handle a command like "State_1".

It just needs "State", and a correct topic. 

I get errors because of it. 

----- Original message -----

**Reply by Koenkk** [reply]
`state_1` doesn't because `1` isn't a valid endpoint name. It will work with `state_l1` when defining the endpoints as `return {'l1': 1, 'l2': 2};`

**Reply by mslotema** [reply]
OMG *facepalm*

It is EL-One. Not "eleven"... I need a different font, or glasses.

Thanks a million!


Still don't fully get the way it is translated, but this works:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const e = exposes.presets;

const device = {
    zigbeeModel: ['Emotion'],
    model: 'Emotion',
    vendor: 'LS Deutschland GmbH',
    description: 'LS Emotion zigbee bridge for LED controllers',
    fromZigbee: [fz.on_off, fz.brightness, fz.color_colortemp],
    toZigbee: [tz.light_onoff_brightness, tz.ignore_transition, tz.ignore_rate, tz.effect,
        tz.light_brightness_move, tz.light_brightness_step, tz.light_colortemp,
        tz.light_colortemp_move, tz.light_colortemp_step],
    meta: {disableDefaultResponse: true, multiEndpoint: false},
    exposes: [
        e.light_brightness_colortemp()
            .removeFeature('color_temp_startup')
            .withEndpoint('l1'),
        e.light_brightness_colortemp()
            .removeFeature('color_temp_startup')
            .withEndpoint('l2'),
    ],
    endpoint: (device) => {
        return { 'l1': 1, 'l2': 2};
    }
};

module.exports = device;
```

Is there a way to examine the number of endpoints it exposes by, e.g. accessing the interview? Pretty sure it should be exposing more endpoints, when I bind it to more of the powerblocks.

**Reply by Koenkk** [reply]
You can look at the database.db file, this will show all endpoints of the device

**Comment by jocelyngirard** [comment]
Hi @mslotema,

Have you successfully added/implemented the L&S Emotion device to zigbee2mqtt ?
I am a new owner of it, trying some configurations, but unable to interact with.

is it possible to add this device to the z2m supported device ?

**Reply by mslotema** [reply]
Hi @jocelyngirard ,

I have it working. But... I think there is still some stuff to work out (dynamically added "zones" to name something - I have 2), before I feel it is good enough for submitting it to the actual z2m repository.

I am happy to share, and write down how to get it working.

Let me know. :-)

Regards,
Melle
